### PR TITLE
Hide debit/credit card SPBs for subscription purchases

### PR DESCRIPTION
**Pull request by james-allan**
GitHub comment: https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/668#issuecomment-604242940

---

### Description

When purchasing a subscription, you cannot use debit or credit card smart payment buttons because we need to create a billing agreement and therefore need a PayPal account. This PR hides credit and debit card fields when viewing a subscription product page or if the cart contains a subscription.

Unfortunately, the changes needed here were more than I would have liked. This is because of the way the `get_button_settings()` function sometimes gets an empty string or an array of hidden methods and so we needed to work around that. 

### Steps to test:

1. Set up your [PayPal Checkout settings](https://user-images.githubusercontent.com/8490476/77714957-0dc05c80-7026-11ea-8d86-a06b9b51a6a8.png) so all payment options are visible.
1. On `master` go to a subscription product's page and try to purchase the product using [one of the credit card options](https://user-images.githubusercontent.com/8490476/77715284-ecac3b80-7026-11ea-92cd-531384b9a9c4.png). You should see an [error](https://user-images.githubusercontent.com/8490476/77715910-96d89300-7028-11ea-81c8-ab558add6dc0.png). 
1. Switch to this branch. The credit card fields will be hidden.
1. You can also test using the old JS Smart payment buttons by removing the REST API Client ID credential.

@jorgeatorres what are your thoughts on this approach? I tried heaps of possible fixes, I tried using a filter but couldn't find a good place to add a callback because this runs before all classes with Subscriptions integration. The current set of changes include some duplicate code, before the SDK code overrides the hidden methods generated by `get_button_settings()`. 

### Documentation

- [x] This PR needs documentation (has the "Documentation" label).

**Note:** 
When using the new JS SDK, if you're viewing a subscription's single product page, you won't be able to see/use the credit card fields in the mini-cart because there's no longer any context-specific hidden methods. 

For example: 
On Subscription page:
Screen Shot 2020-03-27 at 3 28 36 pm

On simple product page:
Screen Shot 2020-03-27 at 3 30 47 pm

### CODE REVIEW

File: includes/class-wc-gateway-ppec-cart-handler.php [hunk 1]
```diff
@@ -415,28 +415,44 @@ public function maybe_enqueue_checkout_js( $widget_title, $widget_instance = arr
...
-				$data['hide_funding_methods'] = array( 'credit' );
-			} elseif ( ! in_array( 'credit', $data['hide_funding_methods'] ) ) {
+			if ( ! in_array( 'credit', $data['hide_funding_methods'] ) ) {
 				$data['hide_funding_methods'][] = 'credit';
 			}
 		}
 
+		// If we need to create a billing agreement or this is a subscription's single product page, hide credit/debit card options.
+		if (
+			! in_array( 'card', $data['hide_funding_methods'] )
```
Comment (jorgeatorres) [inline includes/class-wc-gateway-ppec-cart-handler.php]: While not critical, the condition here is missing some parenthesis: it should be `A && (B || C)` but it currently is `A && B || C` which PHP translates to `(A && B) || C` due to operator precedence.
What this means is that "card" gets added twice to the array if it was already in the array *and* you're on a subscription product page (that is, A = false, C = true).

File: includes/class-wc-gateway-ppec-cart-handler.php [hunk 2]
```diff
@@ -527,6 +543,14 @@ public function enqueue_scripts() {
 					( ! $settings->is_credit_enabled() ? array( 'credit' ) : array() )
 				);
 				$data['disallowed_methods'] = array_map( 'strtolower', $data['disallowed_methods'] );
+
+				if (
+					! in_array( 'card', $data['disallowed_methods'] )
```
Comment (jorgeatorres) [inline includes/class-wc-gateway-ppec-cart-handler.php]: Same as the other condition (with the scenarios reversed). I also think it'd be useful to add the comment about hiding credit/debit card options to clarify what's happening here.

File: includes/class-wc-gateway-ppec-cart-handler.php [hunk 3]
```diff
@@ -415,28 +415,44 @@ public function maybe_enqueue_checkout_js( $widget_title, $widget_instance = arr
...
-			} elseif ( ! in_array( 'credit', $data['hide_funding_methods'] ) ) {
+			if ( ! in_array( 'credit', $data['hide_funding_methods'] ) ) {
 				$data['hide_funding_methods'][] = 'credit';
 			}
 		}
 
+		// If we need to create a billing agreement or this is a subscription's single product page, hide credit/debit card options.
+		if (
+			! in_array( 'card', $data['hide_funding_methods'] )
+			&& ( 'single_product' !== $context && wc_gateway_ppec()->checkout->needs_billing_agreement_creation( array() ) )
```
Comment (jorgeatorres) [inline includes/class-wc-gateway-ppec-cart-handler.php]: Checking that context is "single_product" is not enough to hide the buttons in all cases: PPEC only calls this function (`get_button_settings()`) with "single_product" as context when [you're on a single product page **and** the single product specific settings are enabled](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/3b40bb886be65102dcd09d8436436e1d971ce212/includes/class-wc-gateway-ppec-cart-handler.php#L514-L515).

If you're on a single product page but have no single product specific settings, this function is called with no context (default), but we still need to hide the buttons, which doesn't happen.

**Steps to test**:
1. Make sure "Configure settings specific to Single Product view" is not checked.
2. Visit a single product page.
3. Credit/debit card options are (incorrectly) shown.

- - -

Not sure if this means a filter implementation is more likely to work, as you can't rely on the context alone. Maybe [a filter around _this_ value](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/3b40bb886be65102dcd09d8436436e1d971ce212/includes/class-wc-gateway-ppec-cart-handler.php#L522) could work.

**Comment by jorgeatorres** [comment]
Thanks for working on this @james-allan!

> @jorgeatorres what are your thoughts on this approach? I tried heaps of possible fixes, I tried using a filter but couldn't find a good place to add a callback because this runs before all classes with Subscriptions integration. The current set of changes include some duplicate code, before the SDK code overrides the hidden methods generated by get_button_settings().

I was thinking more of a filter we'd use inside PPEC itself (not Subscriptions) in case that was useful to avoid some duplication. That said, I have no objections against the approach you took here.
I still think we need to refactor those functions (`get_button_settings()` and `enqueue_scripts()`) as making reasonable changes (such as this one) becomes incredibly involved, but right now is probably not the best time to do such refactor.

- - -

I left some review comments. Let me know what you think 😸.

File: includes/class-wc-gateway-ppec-cart-handler.php (hunk 3 above)
Comment (james-allan) [reply includes/class-wc-gateway-ppec-cart-handler.php:445]: > If you're on a single product page but have no single product specific settings, this function is called with no context (default), but we still need to hide the buttons, which doesn't happen.

I don't think the card fields are displayed in this case. 

For example these are my settings. 
Screen Shot 2020-04-03 at 1 46 34 pm

On a subscription page the credit card methods aren't displayed.

Screen Shot 2020-04-03 at 1 47 32 pm

On a simple product they are.

Screen Shot 2020-04-03 at 1 48 37 pm

This is handled by this condition. https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/hide_cards_for_subscriptions/includes/class-wc-gateway-ppec-cart-handler.php#L547-L553

**Comment by james-allan** [comment]
> I was thinking more of a filter we'd use inside PPEC itself (not Subscriptions) in case that was useful to avoid some duplication. 

Yeah sorry. I wasn't clear on that. I just meant that the classes which implement subscription integration in PPEC like `WC_Gateway_PPEC_With_SPB_Addons` are constructed after this function so the filter wouldn't be attached by the time it runs. 

That just means we'd need to attach the filter in another class, which is doable of course but I couldn't find an obvious alternative.

----

Re your comments, I've responded to one of them and fixed the others in e0da0a4. 🙂

File: includes/class-wc-gateway-ppec-cart-handler.php (hunk 3 above)
Comment (jorgeatorres) [reply includes/class-wc-gateway-ppec-cart-handler.php]: Hmm that's weird, that we're seeing different things. This works correctly for me when in SDK mode, but not in legacy mode. In fact, the condition you've mentioned is inside a `$use_js_sdk` so it doesn't apply to legacy mode.

I believe the problem is that in legacy mode we rely entirely on `get_button_settings()` and the "single_product" context, but this function doesn't get called with said context [unless there are single product specific settings](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/e0da0a4896e162c6866417aebd72868a1aeb9bc2/includes/class-wc-gateway-ppec-cart-handler.php#L514-L515) (which is separate from just enabling PPEC on this view).

I'm not sure why it works for you, though. Could you repeat the test again just to confirm?

**Comment by jorgeatorres** [comment]
@james-allan: Thanks for making the changes! 💯 
I'm still having trouble with the card buttons on single product pages. I left another comment, but I'm not sure why we're seeing different things. Could you take another look at it?

File: includes/class-wc-gateway-ppec-cart-handler.php (hunk 3 above)
Comment (james-allan) [reply includes/class-wc-gateway-ppec-cart-handler.php]: ok they have shown up this morning. 🤦‍♂

**Comment by james-allan** [comment]
@jorgeatorres I've made a couple of commits to fix the issue you raised. When I did that I realised I needed to use the exact same condition that I had the main function behind the `$is_sdk` check so I just moved it to run under all cases. 

When I did that I had to make a couple of additional changes. Let me know about this though. 
I also did a whole bunch of testing and the results are below. 

tldr; With the SDK, there are 3 cases where the credit card fields are displayed when they shouldn't. All of them occur when viewing a single product page and it's eligibility overrules the eligibility of what's in the cart. As far as I can see there's no way to get around this since the SDK doesn't allow for different contexts on the same page.

<table>
  <tr>
    <th>Context</th>
    <th>Mode</th>
    <th>Simple Product</th>
    <th>Subscription Product</th>
    <th>Mixed</th>
  </tr>
  <tr>
    <td rowspan="2">Product</td>
    <td>SDK</td>
    <td>Cards shown (<a href="https://d.pr/i/Efvps8">https://d.pr/i/Efvps8</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/WafK3W">https://d.pr/i/WafK3W</a>)</td>
    <td>NA</td>
  </tr>
  <tr>
    <td>Legacy</td>
    <td>Cards shown (<a href="https://d.pr/i/Ib38Xx">https://d.pr/i/Ib38Xx</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/0VzjzL">https://d.pr/i/0VzjzL</a>)</td>
    <td>NA</td>
  </tr>
  <tr>
    <td rowspan="2">Minicart (on simple product)</td>
    <td>SDK</td>
    <td>Cards Shown (<a href="https://d.pr/i/tmuUnS">https://d.pr/i/tmuUnS</a>)</td>
    <td><strong>Cards shown (<a href="https://d.pr/i/bzdvx4">https://d.pr/i/bzdvx4</span></a><span style="font-weight:bold">) 🔴<strong></td>
    <td><strong>Cards shown (<a href="https://d.pr/i/mFCo6w">https://d.pr/i/mFCo6w</span></a>) 🔴</strong></td>
  </tr>
  <tr>
    <td>Legacy</td>
    <td>Cards shown (<a href="https://d.pr/i/PjKNuL">https://d.pr/i/PjKNuL</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/fwVVRq">https://d.pr/i/fwVVRq</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/mzP6TQ">https://d.pr/i/mzP6TQ</a>)</td>
  </tr>
  <tr>
    <td rowspan="2"><span style="font-weight:bold">Minicart (on Subscription product)</span></td>
    <td>SDK</td>
    <td><strong>Cards not shown (</span><a href="https://d.pr/i/9enIoM">https://d.pr/i/9enIoM</span></a>) 🔴</strong></td>
    <td>Cards not shown (<a href="https://d.pr/i/uPxT36">https://d.pr/i/uPxT36</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/WafK3W">https://d.pr/i/WafK3W</a>)</td>
  </tr>
  <tr>
    <td>Legacy</td>
    <td>Cards shown (<a href="https://d.pr/i/gVHPfL">https://d.pr/i/gVHPfL</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/NBLRUc">https://d.pr/i/NBLRUc</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/jwm6hx">https://d.pr/i/jwm6hx</a>)</td>
  </tr>
  <tr>
    <td rowspan="2">Minicart (on shop)</td>
    <td>SDK</td>
    <td>Cards shown (<a href="https://d.pr/i/qNH4sz">https://d.pr/i/qNH4sz</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/1rDgOh">https://d.pr/i/1rDgOh</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/5V6FsS">https://d.pr/i/5V6FsS</a>)</td>
  </tr>
  <tr>
    <td>Legacy</td>
    <td>Cards shown (<a href="https://d.pr/i/Aye3A1">https://d.pr/i/Aye3A1</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/YbNugw">https://d.pr/i/YbNugw</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/bFNpxY">https://d.pr/i/bFNpxY</a>)</td>
  </tr>
  <tr>
    <td rowspan="2">Cart</td>
    <td>SDK</td>
    <td>Cards shown (<a href="https://d.pr/i/PT3g3v">https://d.pr/i/PT3g3v</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/SWQHcX">https://d.pr/i/SWQHcX</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/6Fdq9P">https://d.pr/i/6Fdq9P</a>)</td>
  </tr>
  <tr>
    <td>Legacy</td>
    <td>Cards shown (<a href="https://d.pr/i/nlMX3o">https://d.pr/i/nlMX3o</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/UM3vJL">https://d.pr/i/UM3vJL</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/mht9hj">https://d.pr/i/mht9hj</a>)</td>
  </tr>
  <tr>
    <td rowspan="2">Checkout</td>
    <td>SDK</td>
    <td>Cards shown (<a href="https://d.pr/i/SpprjC">https://d.pr/i/SpprjC</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/1rDgOh">https://d.pr/i/1rDgOh</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/nrFrqr">https://d.pr/i/nrFrqr</a>)</td>
  </tr>
  <tr>
    <td>Legacy</td>
    <td>Cards shown (<a href="https://d.pr/i/KcjVJw">https://d.pr/i/KcjVJw</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/PrgjPP">https://d.pr/i/PrgjPP</a>)</td>
    <td>Cards not shown (<a href="https://d.pr/i/FT97Yq">https://d.pr/i/FT97Yq</a>)</td>
  </tr>
</table>

**Comment by jorgeatorres** [comment]
Hi @james-allan! Thank you for making the changes.

> As far as I can see there's no way to get around this since the SDK doesn't allow for different contexts on the same page.

I found a way to do context specific settings again, and as such, I believe we would be able to make all your test cases pass. That said, I'm still working on the changes. I'm going to update the `update_javascript_sdk` branch as soon as I'm done so that we can revisit this PR. I'll let you know.

**Comment by jorgeatorres** [comment]
@james-allan: `update_javascript_sdk` should be ready now. That said, I made a mess of your branch. I tried to rebase it to solve the conflicts but there were a ton. I think it'd be easier to cherry-pick the latest commits or just copy/paste the code. I'm really sorry 😿.

Based on the new possibilities in `update_javascript_sdk`, I gave the filter approach another go. I'm not sure if it works for all cases (or if a filter is really needed) but it could be used as a starting point. Here's the branch: https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/compare/update_javascript_sdk...hide_cards_for_subscriptions_redux.

I didn't go through all the test cases above, so I might be missing something. Please let me know what your thoughts are. Feel free to adapt it or ignore it entirely (depending on what you find out).

Once again, sorry for making you re-do this PR several times. I didn't know we could do context specific settings on the new SDK until today 🙄.

**Comment by james-allan** [comment]
> Once again, sorry for making you re-do this PR several times. I didn't know we could do context specific settings on the new SDK until today

Haaha no worries @jorgeatorres. It's all good. At least there's a clear way forward now. I definitely prefer this approach because wrangling the SDK vs legacy settings was confusing. 😆

**Comment by james-allan** [comment]
Ok I've essentially started this PR from scratch and went with the filter approach. I've run through all the test cases above and the latest changes ensure the credit card fields are displayed or hidden in all the right places.

**Comment by jorgeatorres** [comment]
@james-allan: Thanks again! I pushed a commit to fix a docblock typo. Besides that, everything looked great. LGTM 🚢.
